driver-core/src/main/com/mongodb/internal/time/Timeout.java

```diff
@@ -13,164 +13,119 @@
+ * for tracking timeouts.
+ * <p>
+ * This class is not part of the public API and may be removed or changed at any time.</p>
  */
 @Immutable
 public final class Timeout {
-    private static final Timeout INFINITE = new Timeout(-1, 0);
-    private static final Timeout IMMEDIATE = new Timeout(0, 0);
+    private static final Timeout INFINITE = new Timeout(-1, Timer.useless());
+    private static final Timeout IMMEDIATE = new Timeout(0, Timer.useless());
 
     private final long durationNanos;
-    private final long startNanos;
+    private final Timer timer;
 
-    private Timeout(final long durationNanos, final long startNanos) {
+    private Timeout(final long durationNanos, final Timer timer) {
         this.durationNanos = durationNanos;
-        this.startNanos = startNanos;
+        this.timer = timer;
     }
 
     /**
-     * Converts the specified {@code duration} from {@code unit}s to {@link TimeUnit#NANOSECONDS} via {@link TimeUnit#toNanos(long)}
-     * and then acts identically to {@link #startNow(long)}.
+     * Converts the specified {@code duration} from {@code unit}s to {@link TimeUnit#NANOSECONDS}
+     * as specified by {@link TimeUnit#toNanos(long)} and then acts identically to {@link #started(long, Timer)}.
      * <p>
      * Note that the contract of this method is also used in some places to specify the behavior of methods that accept
      * {@code (long timeout, TimeUnit unit)}, e.g., {@link com.mongodb.internal.connection.ConcurrentPool#get(long, TimeUnit)},
-     * so it cannot be changed without updating those methods.
-     * @see #startNow(long)
+     * so it cannot be changed without updating those methods.</p>
+     * @see #started(long, Timer)
      */
-    public static Timeout startNow(final long duration, final TimeUnit unit) {
-        assertNotNull(unit);
-        return startNow(unit.toNanos(duration));
+    public static Timeout started(final long duration, final TimeUnit unit, final Timer timer) {
+        return started(unit.toNanos(duration), timer);
     }
 
     /**
      * Returns an {@linkplain #isInfinite() infinite} timeout if {@code durationNanos} is either negative
      * or is equal to {@link Long#MAX_VALUE},
      * an {@linkplain #isImmediate() immediate} timeout if {@code durationNanos} is 0,
-     * otherwise an object that represents the specified {@code durationNanos}.
+     * otherwise a timeout of {@code durationNanos}.
      * <p>
      * Note that the contract of this method is also used in some places to specify the behavior of methods that accept
      * {@code (long timeout, TimeUnit unit)}, e.g., {@link com.mongodb.internal.connection.ConcurrentPool#get(long, TimeUnit)},
-     * so it cannot be changed without updating those methods.
+     * so it cannot be changed without updating those methods.</p>
      */
-    public static Timeout startNow(final long durationNanos) {
+    public static Timeout started(final long durationNanos, final Timer timer) {
```

Comment:
Accepting `Timer` allows

- having multiple timeouts that start at the same instant
- [more importantly for `JAVA-5076`] having a timer and a timeout started at the same instant.

Reply:
I understand that Timeout is a class that can be "started" only at the time that `startNow` is called, and it times out ("expires") at a certain future time. This appears to be adequate for the above requirements, so I am unsure why Timer is needed. The name "Timer" is also confusing... it does not seem to "time" anything, but is a wrapper around a start time, making it seem more like an Instant. It is good that this PR split from the other PR, but since it is split, I may be misunderstanding the requirements:

* Timeout itself should be a value class, so I think this means that you need multiple timeouts with the same startNanos, but different durationNanos. It seems this can be better accomplished by introducing a `withDuration` method, that returns a class with the same startNanos, but a different duration. The implementation of INFINITE and IMMEDIATE should be changed: I do not think that they are correct as they are. It seems they should still record the correct startTime. For example, the isImmediate method should return true when any timer is started with a duration of 0, but it presently checks for equality against IMMEDIATE. I think perhaps 0 and MAX_LONG should be used to represent immediate and infinite, and that there should be no Timeout constants.
* It appears from the other PR that the only reason to have a timer and a timeout started at the same time is to [get the elapsed time.](https://github.com/mongodb/mongo-java-driver/pull/1166/files#diff-7f8dc15248ba3fe10795342e55c0525fee969cc73d918141a5da85dbded487fbR274) I think this can be handled by getting the elapsed time from a Timeout directly. If there is some other requirement, could you please expand on this?

Reply:
There are many things packed in the above comment, and I am not sure most of them leads to something. I tried replying to all of them and grouping replies in numbered sections.

## 1. I am unsure why `Timer` is needed

It is needed for "for tracking elapsed time", as documented. Since it is now needed not only for implementing `Timeout`, but also on its own, the functionality was extracted from internals of `Timeout` into a separate class.

> "Timer" ... is a wrapper around a start time

It is important to point out that the "start time" value is practically meaningless, and only the difference between two such values is meaningful. Both the Go and Rust standard libraries provide tools for measuring elapsed time that hide such meaningless "start time" values. Unfortunately, the Java SE API  does not do that and instead exposes `System.nanoTime`. This is most likely because Java did not have a way to make such an abstraction zero-cost at the time the API was designed (nor does it have such a way now, but it will when value objects and null-restricted types are implemented). The Java SE API does have to be concerned with use cases like microbenchmarks, but we do not, and we can afford to introduce the abstraction.

> ... need multiple timeouts with the same startNanos, but different durationNanos. It seems this can be better accomplished by introducing a `withDuration` method, ...

This is an alternative approach to the current API (it also eliminates the need to have the `duration` parameter in the `Timeout.started` methods), but I don't think it is better.

> ... get the elapsed time. I think this can be handled by getting the elapsed time from a Timeout directly.

It is true that technically the `Timeout` class could provide API for measuring elapsed time, it could similarly provide any other API that is not needed to achieve the purpose of `Timeout`. But I think that it is better to have a separate tool for this job.

Just in case you are surprised that given the above, there are `Timeout.remaining`/`remainingOrInfinite` methods - they are an unfortunate byproduct of us having to integrate with external APIs that accept a timeout in a form different from our `Timeout`. While an approach that does not expose `Timeout.remaining`/`remainingOrInfinite` methods is possible, it requires separate `Timeout` methods for integrating with each external API, and results in code that does not look like usual code that uses those external APIs.

## 2. The name "Timer" is also confusing... it does not seem to "time" anything

I used this name because this is how you called a [similar class](https://github.com/mongodb/mongo-kafka/commit/c25e0526bfd43c619b2e06a2739b7c525a7af9ec#diff-136a2047e289c24f10ac62956d07711230e8e438a2680d9be51fbfc4770b58f4R19) in MongoDB Kafka connector, so I am surprised there is now a concern about the name. However, **I agree that it is not a "timer" but a stopwatch, and I am happy to rename it this way if you agree.**

## 3. `Timer` seem more like an `Instant`

Seeing `Timer` not as a stopwatch, but as a point on a timeline with the origin somewhere within the time interval in which the `Timer.start` method was running, is a valid way of interpreting it. In fact, both the Go and Rust standard libraries do exactly this: in Go the type that allows measuring elapsed time is called `Time`<sup>1</sup>, and in Rust the corresponding struct is called `Instant` (despite having the same name, it is not semantically similar to `Instant` from the Java SE API). **I will be happy to use this interpretation, if this is what you propose** (this approach in conflict with item 2 above: only one may be chosen). **If so, do you have a name for the class better than `Time`?**

## The implementation of INFINITE and IMMEDIATE should be changed: I do not think that they are correct

> For example, the isImmediate method should return true when any timer is started with a duration of 0

What you described, is how `isImmediate` behaves. See `assertEquals(Timeout.immediate(), Timeout.started(0, timer)` in `TimeoutTest.started`.

## 4. 0 and MAX_LONG should be used to represent immediate and infinite

0 does represent an immediate timeout, and `Long.MAX_VALUE` does represent an infinite one.

## 5. there should be no Timeout constants

This PR does not introduce the `Timeout.immediate`/`infinite` method, but I know why having the them is convenient: they are useful in situations when an immediate/infinite timeout is statically known to be needed; for `Timeout.immediate()` we already have such a place in the codebase. I don't know what substantiates the opinion that `Timeout.immediate`/`infinite` methods should not exist.

---

<sup>1</sup> `Time` in the Go standard library combines the functionality of a monotonic clock with the functionality of a wall clock (a.k.a. system clock). I neither understand the reasoning behind such an API design, nor how the API is implemented.

Reply:
I have realized that the interpretation I described in section 3

> ... a point on a timeline with the origin somewhere within the time interval in which the `Timer.start` method was running, ...

was weaker than what we have in reality, as it did not allow comparing different points, while we can, in fact, compare them. I updated it by making the description stronger, such that it allows for point comparison.

Reply:
After thinking more about the interpretation in section 2 vs. interpretation in section 3, I find the latter more useful and convenient, so I will just make changes accordingly.

Reply:
Of all the standard libraries for languages for which there are official MongoDB drivers, only the following provide an abstraction (either the one described in section 2, or in section 3 above):

| Language | Relevant type name and link | Abstraction, section 2 or 3 |
| --- | --- | --- |
| C# | [`Stopwatch`](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.stopwatch?view=net-7.0) | section 2 |
| C++ | [`time_point`](https://en.cppreference.com/w/cpp/chrono/steady_clock) | section 3 |
| Kotlin | [`ValueTimeMark`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.time/-time-source/-monotonic/-value-time-mark/) | section 3 |
| Swift | [`Instant`](https://developer.apple.com/documentation/swift/continuousclock#structures) | section 3 |
| Go | [`Time`](https://pkg.go.dev/time#Time) | section 3 |
| Rust | [`Instant`](https://doc.rust-lang.org/std/time/struct.Instant.html) | section 3 |

The approach described in section 3 is much more popular, with `Instant` being the most popular name. Since it is already taken by a class in the Java SE API, I picked `TimePoint` from C++. `Time` from Go does not feel specific enough, and `ValueTimeMark` form Kotlin feels like a fruit of a diseased imagination.

Reply:
This PR sets out how we should time things in the driver, and will affect more than the linked PR. We seem to have the following requirements:

* Determine how much time has `elapsed` since a certain past time
* Determine whether an originally-future point in time has passed (timed out, `expired`)...
* ...and also how much time is `remaining` before that point (it seems unimportant whether arises from an internal or external need, it's still something we must support and represent)

Currently, `Timeout` covers `expired`/`remaining`. This PR currently proposes `TimePoint`, to cover `elapsed`. This introduces a new class, and an interaction with the existing class (Timeouts must be started using TimePoints). If we take this approach, I think that we should preserve the existing `startNow` methods. They are at least fine convenience methods, and in any case there seems to be no need to couple the classes together in this way.

A Timeout appears to be just a TimePoint that is originally intended to be in the future, though it might eventually (or immediately) be in the past. It seems unnecessary to implement this same concept using two different classes. This seems to be the approach taken by Kotlin's ValueTimeMark. If we have a strong justification for needing to differentiate between a TimePoint that tracks elapsed time, and a TimePoint that tracks remaining time, we could use interfaces to hide the appropriate methods.

Reply:
I think that the idea to replace timeouts with deadlines, which can be represented as `TimePoint`s is really great!👌

- We will need `TimePoint.infinity()` for the deadline corresponding to `Timeout.infinite()`, and it can be encoded with `TimePoint.nanos` being `null`.
- Despite there being no special value left (`null` is the only special value we have in this approach) for encoding a ingle deadline for `Timeout.immediate()`, every result of `TimePoint.now()` is such a deadline.

I'll implement this approach here.

Reply:
> I'll implement this approach here.

I jumped the gun there. After discussing with Jeff whether I should do such refactoring now, or let the team triage the idea, I created https://jira.mongodb.org/browse/JAVA-5112. So if we start using deadlines instead of timeouts internally, that will be a PR for the created task.

Reply:
> we should preserve the existing `startNow` methods

Done.